Added Zoe intro, updated Doc Guide and menus (#175)
Agoric/documentation: main/getting-started/intro-zoe.md

For example, let’s say a realtor has a standard agreement they use when selling a client’s house. The contract would be the code that defines how that standard agreement works. Every time the realtor has a new house to sell, they instantiate a new instance of their standard contract to use for that specific property. If they have ten houses up for sale, they have ten different contract instances.

In addition to an `installationHandle`, when the contract is instantiated, Zoe needs to know what kind of ERTP assets the contract will work with. In Atomic Swap’s test code, the full instantiation call is:

```js
const issuerKeywordRecord = harden({
  Asset: moolaIssuer,
  Price: simoleanIssuer,
});
const aliceInvite = await E(zoe).makeInstance(
  installationHandle,
  issuerKeywordRecord, 
});
```

> Let’s work backwards from `makeInstance()`’s arguments. As you’ll recall from the [Introduction to ERTP](https://agoric.com/documentation/getting-started/ertp-introduction.html#issuers-and-mints), *issuers* map minted assets to purses and payments. The *keyword record* is made up of two key:value pairs. The key names must be ASCII and capitalized; they are the same as are used in the contract code, in this case `Price` and `Asset`. `Price` is for what is wanted, and `Asset` is for what is being sold. `Price` is for what will be swapped for it (from the perspective of the user making the instance; it’d be the opposite for someone who wanted what Alice is offering). The price is denominated in simoleons, so that keyword needs the simoleon Issuer associated with it. The asset is denominated in moola, so that keyword needs the moola Issuer as its value. 

Reviewer: > `Price` is for what is wanted, and `Asset` is for what is being sold. `Price` is for what will be swapped for it 

`Asset` is for what is being sold. `Price` is for what will be swapped for it 

I'd drop the first clause. The second description of Price is clearer and sufficient

> Let’s work backwards from `makeInstance()`’s arguments. As you’ll recall from the [Introduction to ERTP](https://agoric.com/documentation/getting-started/ertp-introduction.html#issuers-and-mints), *issuers* map minted assets to purses and payments. The *keyword record* is made up of two key:value pairs. The key names must be ASCII and capitalized; they are the same as are used in the contract code, in this case `Price` and `Asset`. `Price` is for what is wanted, and `Asset` is for what is being sold. `Price` is for what will be swapped for it (from the perspective of the user making the instance; it’d be the opposite for someone who wanted what Alice is offering). The price is denominated in simoleons, so that keyword needs the simoleon Issuer associated with it. The asset is denominated in moola, so that keyword needs the moola Issuer as its value. 

Reviewer: > The price is denominated in simoleons, so that keyword needs the simoleon Issuer associated with it. The asset is denominated in moola, so that keyword needs the moola Issuer as its value. 

I'd say explicitly that the currencies are what is being customized here. Every instantiation is going to need a call to makeInstance to specify the currencies. It may be that the real estate agent always trades property for dollars, but a stock broker might use the same contract to trade shares for euros.

Finally, `E(zoe).makeInstance()` can take a `terms` argument, another set of key:value pairs. Terms let a contract instance creator further customize the contract operations, as enabled by the contract code. Contract terms can be expressed without specific values. For example, an auction contract may define minimum bid and minimum raise variables and use them in its code, but the variables are not given values in the code. Instead, the `terms` argument provides the variables’ values.

> Why the need to parameterize a contract with terms? Well, the minimum raise ]should be quite different if an auction is for a used Harry Potter paperback vs. a signed first British hardback edition of *Harry Potter and the Philosopher’s Stone*]. The `terms` value for the used paperback auction would be `{ Currency: 'USDollar', MinRaise: 1 }`. For the first edition auction, `terms`’ value would be  `{ Currency: 'USDollar', MinRaise: 500)`. With the different `terms` values, there are now contract instances customized for the items being sold in each one.

Reviewer: mismatched square brackets.

And while whoever instantiated the contract gets the initial invite to that instantiation, as with all invites, they can transfer it to another party before it’s used. Who created or first held an invite doesn’t matter. Whoever possesses an invite has control and use of it. 

Users obtain invites to a contract instance in one (or more) of three ways:
- Instantiate a contract, which returns an invite object for that contract instance.
- By receiving an invite from someone who already had it. 
  - Note: Finding other parties to invite is out of Zoe or the contract’s scope in most cases. You have to either already know of potential other parties, for example, from past experience trading with them, or communicate somewhere outside the contract to find them.
- By the contract instance or someone with the authority to issue invites for this contract instantiation making open knowledge what’s needed to get a valid invite. For example, an open exchange contract with an order book might welcome everyone who wants to participate, and so would make public how to get an invite to it.

> *Invites* are ERTP *payments*, so the same principles apply to them as apply to other ERTP payments. You can send the same payment or invite to multiple parties...but when a party claims the payment, that payment is dead. The claimant gets a new payment they know was issued by the desired issuer, and which they have exclusive access to. 

Reviewer 2: maybe instead: "but when a party claims the payment after someone else, that payment is dead"

The offers are now completed, and nothing more is or can be done with them. Zoe deletes completed offers from the contract instance.

## Auction example

Let’s look at a more complicated example: an auction, where many users might make a bid offer for something, but only one offer will win the auction. Assuming everything validated, Zoe would give the holder of the auctioned item the escrowed amount from the winning offer, and give the holder of the winning offer what they wanted, the item that had been up for bid.

But what about all those other bidders who escrowed their bid amounts with Zoe and didn’t win? Zoe guarantees they all get a *refund* of their escrowed amount. Zoe’s *offer safety* guarantees offer makers either get what they wanted or get refunded whatever they put in Zoe’s escrow with their offer.

> It’s even possible to get both what you wanted in an offer, and at least a partial refund of what you offered. Consider an auction where you make a bid offer where you’re willing to pay up to 10 quatloos, but if you can win with a lower bid, that’s your offer. You win the auction, but your winning bid amount is only 8 quatloos. The payout would resolve to both the item up for auction you get as the winning bidder and the refund of the 2 extra quatloos you escrowed in the offer. 

Reviewer: Perhaps not all readers are used to this idea, so we should make it explicit:

"Consider an auction where your bid specifies your maximum payment, but the auctioneer charges just enough to beat the next highest bidder."  Or you could refer to a "second price auction", in which the winner pays the amount bid by the second highest bid. This is also known as a "Vickrey Auction", and you could link to wikipedia for an explanation.

## Other things to know about Zoe

A *Dapp (distributed application)* is a combination of a Zoe contract and a server that’s running the back and front ends, and a front end that may be connected to the user’s wallet.

Zoe has two distinct APIs. One is the widely accessible, singular Zoe service object that smart contract clients use to provide offers and otherwise interface with contracts. The Zoe service object is a remote object, and so should be accessed with `E(zoe).<api_method>`, which returns a promise. For more on E() see [here](https://agoric.com/documentation/distributed-programming.html#communicating-with-remote-objects-via-e). 

> The other is the internal API object, referred to as `zcf` that each contract instance uses to interact with Zoe about that contract’s own offers and state (e.g., to do `reallocate` and `complete`).

Reviewer: "The other is the API for Zoe's Contract Facet, referred to as `zcf`, that each contract instance uses to"

## Workflow summary

> The following table summarizes an Atomic Swap workflow:

Reviewer: ```suggestion
The following table summarizes the workflow for developing a smart contract like Atomic Swap:
```

The flow after instantiation and invitation would be different for other contracts.

<table>
<tbody>
<tr>
<td><center>1</center></td>
<td>Install Zoe on a server and start it running. 
</td>
</tr>
<tr>
<td><center>2</center></td>
<td>Write your contract
</td>
</tr>
<tr>
<td><center>3</center></td>
<td>Install your contract’s source code on Zoe. 
</td>
</tr>
<tr>
<td><center>4</center></td>
<td>Start a contract instance based on an installed contract. 
</td>
</tr>
<tr>

> <td colspan="2">At this point, everything is installed and set up and you’ve got a contract instance. The following rows covers what happens to offers in a contract instance.

Reviewer 2: ```suggestion
<td colspan="2">At this point, everything is installed and set up and you’ve got a contract instance. The following rows cover what happens to offers in a contract instance.
```

</td>
<tr>
<tr>
<td><center>1</center></td>
<td>A holder of an invite to this contract instance uses it to make an offer. The assets used in the offer are escrowed with Zoe.
</td>
</tr>
<tr>
<td><center>2</center></td>

> <td>Likely via means outside of the contract, invites are distributed to potential users of this contract instance).</td>

Reviewer 2: ```suggestion
<td>Likely via means outside of the contract, invites are distributed to potential users of this contract instance.</td>
```
